### PR TITLE
feat: add devcontainer feature for radcli

### LINKED ISSUE #6906: Radius CLI as Devcontainer feature
## Overview of feature request

Proivde the Radius CLI as a devcontainer feature (see <https://containers.dev/features>) to make the provisioning of a devcontainer environment easier. 

## Acceptance criteria

The radius CLI is available as devcontainer feature in the official and community-supported devcontainer feature page (<<https://containers.dev/features>>) and can be used within dev containers via the `feature` section

## Additional context

Happy to suport on this topic

[AB#10644](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/10644)

### PR DESCRIPTION
**Pull request by lechnerc77**
# Description

This PR introduces the setup to provision the `radcli` as a devcontainer feature. This comprises the basic feature setup in package `ext` as well as the relevant GitHub Action workflows to validate the feature schema, to test the feature and to publish the feature in the GH packages of the repository.

Points to discuss/for the review:

- Is location of sources in line with the project guidelines?
- Is workflow for validation really needed?

Not part of PR: To make the feature "officially"available it needs to be registered on the devcontainer feature page https://containers.dev/features. This shoudl probably be done by the Radius maintainers once this PR is merged.

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #6906 

### Summary
🚀🧪🛠️

This pull request adds a new dev container feature for the Radius CLI and the necessary workflows to test, validate, and release it. The feature allows users to install and use the Radius CLI in their dev containers with different versions. The workflows use the devcontainers action and the devcontainer CLI to automate the feature lifecycle. The pull request also includes documentation and test scripts for the feature.

> _`devcontainer` features_
> _GitHub workflows to test, deploy_
> _Autumn leaves falling_

### Walkthrough
*  Add and configure three GitHub workflows to test, validate, and release dev container features ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-786e5036d5206693ef00172e6655e7b814833b8a1117ffed490dd9178f7e2d77R1-R31), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-69fccd49f34279d7919fb3ee4f0b661cc4cf710d4f3e909dad216144513685c4R1-R60), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-61a939772971878380ce0f0fec78dcff2159a3db38a1642817be0e426c28716bR1-R21))
  * The `devcontainer-feature-release.yaml` workflow publishes the features from the `ext/devcontainer/src` directory to the devcontainers registry using the devcontainers action ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-786e5036d5206693ef00172e6655e7b814833b8a1117ffed490dd9178f7e2d77R1-R31))
  * The `devcontainer-feature-test.yaml` workflow tests the features against different base images and scenarios using the devcontainer CLI ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-69fccd49f34279d7919fb3ee4f0b661cc4cf710d4f3e909dad216144513685c4R1-R60))
  * The `devcontainer-feature-validate.yaml` workflow validates the `devcontainer-feature.json` files for each feature using the devcontainers action ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-61a939772971878380ce0f0fec78dcff2159a3db38a1642817be0e426c28716bR1-R21))
* Add a new dev container feature for the Radius CLI ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-6a82e3d8f4cc1c53e79c9d811863da1c8e6c560eaf5399a69888cf34e3b54e15R1-R29), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-b6696c2316f872901334f316b093d203dcb4052b98d0f049bfa248b83ec31c94L1-R17), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-e79890431f6a5417ac328841d3e716da24216d75bb60bf653cfebf43d3554c28R1-R31), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-948172e104408d7b88023a905d4c9e8536c3ed0e76e01b0b9313f6e5bf431787L1-R12), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-cc69f6dfc7cf74da17d60297adb1fe30c66e7eb990266c9fa3b9b7ca21d6b889L1-R9), [link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-4e7f6e849ae923ba58b75122a7d76b9ae78fee0735d4fe083ee463efc6772510L1-R13))
  * The `devcontainer-feature.json` file defines the name, id, version, description, documentationURL, and options of the feature ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-b6696c2316f872901334f316b093d203dcb4052b98d0f049bfa248b83ec31c94L1-R17))
  * The `install.sh` file installs the Radius CLI and its dependencies on the dev container ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-e79890431f6a5417ac328841d3e716da24216d75bb60bf653cfebf43d3554c28R1-R31))
  * The `README.md` file provides a brief description, example usage, and a table of options for the feature ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-6a82e3d8f4cc1c53e79c9d811863da1c8e6c560eaf5399a69888cf34e3b54e15R1-R29))
  * The `test.sh` file tests the feature for the default or autogenerated scenarios using the `dev-container-features-test-lib` library ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-4e7f6e849ae923ba58b75122a7d76b9ae78fee0735d4fe083ee463efc6772510L1-R13))
  * The `edge.sh` file tests the feature when the `version` option is set to `edge` using the same library ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-948172e104408d7b88023a905d4c9e8536c3ed0e76e01b0b9313f6e5bf431787L1-R12))
  * The `scenarios.json` file defines a scenario named `edge` that uses the `mcr.microsoft.com/devcontainers/base:ubuntu` image and sets the `version` option to `edge` for the feature ([link](https://github.com/radius-project/radius/pull/6930/files?diff=unified&w=0#diff-cc69f6dfc7cf74da17d60297adb1fe30c66e7eb990266c9fa3b9b7ca21d6b889L1-R9))

### CODE REVIEW

**Comment by rynowak** [comment]
> Is location of sources in line with the project guidelines?

Can I suggest `/main/deploy/devcontainer-feature`? `deploy` isn't the clearest term, but it's where we put existing "installers" like Helm, or the download scripts. 

@radius-project/maintainers-radius feel free to chime in.

File: .github/workflows/devcontainer-feature-release.yaml
Comment (rynowak) [inline .github/workflows/devcontainer-feature-release.yaml:14]:
```suggestion
    if: ${{ github.ref == 'refs/heads/main' && github.repository == 'radius-project/radius' }}
```

We special-case these for our repo so they don't run on forks and create a bunch of spam.

File: .github/workflows/devcontainer-feature-release.yaml
Comment (rynowak) [inline .github/workflows/devcontainer-feature-release.yaml:28]: Do you know how versioning works for this?  

/cc @willdavsmith

File: .github/workflows/devcontainer-feature-release.yaml
Comment (rynowak) [inline .github/workflows/devcontainer-feature-release.yaml:5]: @willdavsmith - what's the right trigger to put here

File: .github/workflows/devcontainer-feature-test.yaml
Comment (rynowak) [inline .github/workflows/devcontainer-feature-test.yaml:51]: Do you know why this is using `continue-on-error` so liberally? I'd want this to fail if the tests fail.

File: ext/devcontainer/src/radcli/devcontainer-feature.json
Comment (rynowak) [inline ext/devcontainer/src/radcli/devcontainer-feature.json:4]: @willdavsmith - it looks like this file hardcodes versions. This will be another thing for us to automate as part of the release.

Comment (rynowak) [reply ext/devcontainer/src/radcli/devcontainer-feature.json:4]: I don't want this to block the PR, but it's something to consider.

File: ext/devcontainer/src/radcli/install.sh
Comment (rynowak) [inline ext/devcontainer/src/radcli/install.sh:30]: @willdavsmith can you help with the right instructions for edge builds, I know you were just working on this.

**Comment by rynowak** [comment]
> Is workflow for validation really needed?

I personally think this sounds really useful. We've historically had problems testing installation, it's just annoying to test. If this makes it easy for us to test then I'm in favor. 

If it's a lot of work to enable let us know and we can help.

File: ext/devcontainer/src/radcli/devcontainer-feature.json
Comment (lechnerc77) [reply ext/devcontainer/src/radcli/devcontainer-feature.json:15]: Yes sure, will adjust the code and update the PR

File: .github/workflows/devcontainer-feature-test.yaml
Comment (lechnerc77) [reply .github/workflows/devcontainer-feature-test.yaml:51]: I have to check, because I saw failures of the run despite this setting when I did my first tries in the local repo. I will check and report back and in case  adjust the PR accordingly

File: .github/workflows/devcontainer-feature-release.yaml
Comment (lechnerc77) [reply .github/workflows/devcontainer-feature-release.yaml:28]: Besides using semantic versioning I did not see any consistent management of versions (yet). As the versioning of the feature  decoupled from the `rad` CLI versioning I think we can stick to changing versions in accordance to changing code that is specific to the feature *per se*

File: .github/workflows/devcontainer-feature-test.yaml
Comment (lechnerc77) [reply .github/workflows/devcontainer-feature-test.yaml:51]: Checked it in my playground repo. This is indeed confusing. I removed the continue-on-error as this clearly shows if something failed on the overview page. Otherwise you must navigate into the run.

I also removed the `test-global` section as we do not have mutiple features that we want to test in one run, but can keep it simple with the feature-specific test parts

File: .github/workflows/devcontainer-feature-release.yaml
Comment (rynowak) [reply .github/workflows/devcontainer-feature-release.yaml:28]: Let me suggest then that we don't try to solve versioning and publishing of the devcontainer as part of your PR. I'll resolve all of the comments I left that are related to that.

The maintainers need to decide how versioning/publishing will work because we'll be handling it as part of each release (or some other cadence). I don't want to block your PR while we figure this out.

File: .github/workflows/devcontainer-feature-release.yaml
Comment (rynowak) [reply .github/workflows/devcontainer-feature-release.yaml:5]: Per other discussion, let's resolve this after the PR goes in.

File: .github/workflows/devcontainer-feature-test.yaml
Comment (rynowak) [inline .github/workflows/devcontainer-feature-test.yaml:10]: Did you see this comment?: https://github.com/radius-project/radius/pull/6930#issuecomment-1845850076

I'd like to put these files in `deploy/devcontainer-feature`

File: ext/devcontainer/src/radcli/README.md
Comment (rynowak) [inline ext/devcontainer/src/radcli/README.md:6]:
```suggestion

This will install the latest stable release of the `rad` CLI.
```

File: ext/devcontainer/src/radcli/README.md
Comment (rynowak) [inline ext/devcontainer/src/radcli/README.md:16]:
```suggestion

This will install the newest available unstable build of the `rad` CLI.
```

File: ext/devcontainer/src/radcli/devcontainer-feature.json
Comment (rynowak) [inline ext/devcontainer/src/radcli/devcontainer-feature.json:15]: 👍 great

File: ext/devcontainer/src/radcli/README.md
Comment (rynowak) [inline ext/devcontainer/src/radcli/README.md:14]: ## Example Usage - Install specific release

This will install the specified release of the `rad` CLI. In this example the release version is `0.28.0`.

```json
"features": {
    "ghcr.io/devcontainers/radius/radiuscli:latest": {
        "version": "0.28.0"
    }
}
```

**Review by rynowak — CHANGES_REQUESTED** [review]
This is looking good overall. Two issues I'd like to resolve:

- Location of the files in the repo (see comments).
- Script for for `edge` install doesn't look right. https://github.com/radius-project/radius/pull/6930/files#r1419428131 paging @willdavsmith

File: ext/devcontainer/src/radcli/install.sh
Comment (willdavsmith) [reply ext/devcontainer/src/radcli/install.sh:29]: sure - we'll easily be able to get the dev builds after this PR is merged: https://github.com/radius-project/radius/pull/6899. Should be within the hour

File: ext/devcontainer/src/radcli/install.sh
Comment (willdavsmith) [inline ext/devcontainer/src/radcli/install.sh:28]:
```suggestion
    wget -q "https://raw.githubusercontent.com/radius-project/radius/main/deploy/install.sh" -O - | /bin/bash -s edge
```

File: .github/workflows/devcontainer-feature-test.yaml
Comment (rynowak) [reply .github/workflows/devcontainer-feature-test.yaml]: Thanks, I feel better about seeing this without `continue-on-error`.

**Review by willdavsmith — DISMISSED** [review]
awesome initiative @lechnerc77 , thanks so much :) I've added instructions for installing the rad CLI edge version

File: .github/workflows/devcontainer-feature-test.yaml
Comment (lechnerc77) [reply .github/workflows/devcontainer-feature-test.yaml:10]: Totally missed that one. Moved the files and synched the references from the workflows

**Comment by lechnerc77** [comment]
@rynowak @willdavsmith : I updated the PR and also tried to make the wording in documentation overall consistent. The enhanced install script made the feature install script a bit cleaner. 
However, there is now one issue with the installation of the *edge* version: this version has a special prerequisite that the stable versions do not have, namely the `oras` CLI.

We can include this dependency via the `dependsOn` property in the feature metadata (see https://github.com/devcontainers/cli/blob/main/CHANGELOG.md#0440 and https://github.com/devcontainers/spec/blob/main/docs/specs/feature-dependencies.md). This property is still in preview. 

Consequences: 
- The `oras` CLI is also installed for stable releases. *I do not know if this has side effects.*
- The workflow for the metadata validation breaks as its validation is based on the stable metadata schema without preview features. So as long as this is on preview we must deactivate or remove this workflow.

Other option:
- Do not use the `dependsOn` property in the metadata.
- Include the `installsAfter` property (see https://containers.dev/implementors/features/#installsAfter) in the feature metadata referencing the oras feature. This will not automatically resolve the dependency.
- Remove the test of the feature for the *edge* installation (test scenario for edage breaks without `oras` CLI) and add documentation for the special setup for using the devcontainer feature with edge to highlight the dependency to the [oras-cli feature](https://github.com/dhoeric/features/tree/main/src/oras).

**Comment by lechnerc77** [comment]
> Code changes look good to me! There's
> 
> Do you know why this new workflow would be failing? https://github.com/radius-project/radius/actions/runs/7157177190/job/19487724159?pr=6930 It looks like `dependsOn` may not be valid: https://containers.dev/implementors/features/

@rynowak Yes you are right with the impact of the property on the validation run. I mentioned that in my last [comment](https://github.com/radius-project/radius/pull/6930#issuecomment-1848925791).  I had to bring in that property to model the dependency to the `oras` CLI for the edge version. The property is still in preview and that is the reason why the validation fails. The test sceanrios work. As mentioned in the comment I see two options to handle this and both come with trade-offs.

**Comment by rynowak** [comment]
> I mentioned that in my last https://github.com/radius-project/radius/pull/6930#issuecomment-1848925791. I had to bring in that property to model the dependency to the oras CLI for the edge version.

Sorry, I missed the comment where you explained all of this. Thanks for sharing the extra info 👍 

> The workflow for the metadata validation breaks as its validation is based on the stable metadata schema without preview features. So as long as this is on preview we must deactivate or remove this workflow.

Based on the tradeoffs you described, I think disabling the metadata validation workflow makes the most sense for right now. 

- I want to make sure 'edge' works.
- I think the tests have value. Historically it's been tough for us to test installation.

It will be easy for us to re-enable the validation in the future when `dependsOn` is no longer a preview feature. 

Does that sound right? happy to chat more if you have concerns I didn't answer.

**Comment by lechnerc77** [comment]
> > I mentioned that in my last [#6930 (comment)](https://github.com/radius-project/radius/pull/6930#issuecomment-1848925791). I had to bring in that property to model the dependency to the oras CLI for the edge version.
> 
> Sorry, I missed the comment where you explained all of this. Thanks for sharing the extra info 👍
> 
> > The workflow for the metadata validation breaks as its validation is based on the stable metadata schema without preview features. So as long as this is on preview we must deactivate or remove this workflow.
> 
> Based on the tradeoffs you described, I think disabling the metadata validation workflow makes the most sense for right now.
> 
> * I want to make sure 'edge' works.
> * I think the tests have value. Historically it's been tough for us to test installation.
> 
> It will be easy for us to re-enable the validation in the future when `dependsOn` is no longer a preview feature.
> 
> Does that sound right? happy to chat more if you have concerns I didn't answer.

@rynowak I fully agree - the tests add more value than the schema validation and if the tests run, the schema validity is basically proven. 
I removed the automatic triggers from the validation workflow and left the `workflow_dispatch`. So a manual test is still possible if desired. Anything else I should add/change or are we good to go?

**Review by rynowak — APPROVED** [review]
:shipit:! Approved. Thanks @lechnerc77

**Comment by rynowak** [comment]
I'll go through the follow-up steps this morning to get this published.

**Pull request merged**
